source/particles/particle.cc

```diff
@@ -0,0 +1,267 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/particles/particle.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Particles
+{
+  template <int dim, int spacedim, typename PropertyType>
+  Particle<dim,spacedim,PropertyType>::Particle ()
+    :
+    location (),
+    reference_location(),
+    id (0),
+    property_pool(NULL),
+    properties(PropertyPool<PropertyType>::invalid_handle)
+  {
+  }
+
+
+  template <int dim, int spacedim, typename PropertyType>
+  Particle<dim,spacedim,PropertyType>::Particle (const Point<spacedim> &location,
+                                                 const Point<dim> &reference_location,
+                                                 const types::particle_index id)
+    :
+    location (location),
+    reference_location (reference_location),
+    id (id),
+    property_pool(NULL),
+    properties (PropertyPool<PropertyType>::invalid_handle)
+  {
+  }
+
+  template <int dim, int spacedim, typename PropertyType>
+  Particle<dim,spacedim,PropertyType>::Particle (const Particle<dim,spacedim,PropertyType> &particle)
+    :
+    location (particle.get_location()),
+    reference_location (particle.get_reference_location()),
+    id (particle.get_id()),
+    property_pool(particle.property_pool),
+    properties ((property_pool != NULL) ? property_pool->allocate_properties_array() : PropertyPool<PropertyType>::invalid_handle)
+  {
+    if (property_pool != NULL)
+      {
+        const ArrayView<PropertyType> my_properties = property_pool->get_properties(properties);
+
+        if (my_properties.size() != 0)
+          {
+            const ArrayView<const PropertyType> their_properties = particle.get_properties();
+
+            std::copy(&their_properties[0],&their_properties[0]+their_properties.size(),&my_properties[0]);
+          }
+      }
+  }
+
+
+  template <int dim, int spacedim, typename PropertyType>
+  Particle<dim,spacedim,PropertyType>::Particle (const void *&data,
+                                                 PropertyPool<PropertyType> &new_property_pool)
+  {
+    const types::particle_index *id_data = static_cast<const types::particle_index *> (data);
+    id = *id_data++;
+    const double *pdata = reinterpret_cast<const double *> (id_data);
+
```

Comment:
Shouldn't these pointers be PropertyType pointers? Maybe I'm missing the point...